### PR TITLE
Fix bug in gaussian_nll_loss

### LINKED ISSUE #53964: GaussianNLLLoss doesn't support the usual reduction='none'
The new Gaussian NLL Loss behaves differently from the other losses by not supporting the usual `none` mode. Instead, it still does a reduction over all dimensions except batch.

What I expect: `gaussian_nll_loss(..., reduction='none')` should return a tensor with the same shape as input, target, and var. The loss is essentially just `0.5 * (torch.log(var) + (input - target)**2 / var)`, which is what I want returned.

What happens: a scalar or a tensor of shape `(N,)` is returned. The implementation does `.view(input.size(0), -1).sum(dim=1)`

Why this matters: I use `reduction='none'` for custom masking, weighing, etc.

P.S. gaussian_nll_loss is missing from the nn.functional documentation.

cc @albanD @mruberry @jbschlosser

### PR DESCRIPTION
**Pull request by nailimixaM**
Fixes #53964. cc @albanD @almson

## Major changes:
- Overhauled the actual loss calculation so that the shapes are now correct (in functional.py)
- added the missing doc in nn.functional.rst

## Minor changes (in functional.py):
- I removed the previous check on whether input and target were the same shape. This is to allow for broadcasting, say when you have 10 predictions that all have the same target. 
- I added some comments to explain each shape check in detail. Let me know if these should be shortened/cut.

Screenshots of updated docs attached. 
Let me know what you think, thanks!

## Edit: Description of change of behaviour (affecting BC):
The backwards-compatibility is only affected for the `reduction='none'` mode. This was the source of the bug. For tensors with size (N, D), the old returned loss had size (N), as incorrect summation was happening. It will now have size (N, D) as expected.

### Example
Define input tensors, all with size (2, 3).
```
input = torch.tensor([[0., 1., 3.], [2., 4., 0.]], requires_grad=True)
target = torch.tensor([[1., 4., 2.], [-1., 2., 3.]])
var = 2*torch.ones(size=(2, 3), requires_grad=True)
```

Initialise loss with reduction mode 'none'. We expect the returned loss to have the same size as the input tensors, (2, 3).
```
loss = torch.nn.GaussianNLLLoss(reduction='none')
```

Old behaviour:
```
print(loss(input, target, var)) 
# Gives tensor([3.7897, 6.5397], grad_fn=<MulBackward0>. This has size (2).
```

New behaviour:
```
print(loss(input, target, var)) 
# Gives tensor([[0.5966, 2.5966, 0.5966], [2.5966, 1.3466, 2.5966]], grad_fn=<MulBackward0>)
# This has the expected size, (2, 3).
```

To recover the old behaviour, sum along all dimensions except for the 0th:
```
print(loss(input, target, var).sum(dim=1))
# Gives tensor([3.7897, 6.5397], grad_fn=<SumBackward1>.
```

doc1
doc2

### CODE REVIEW

File: torch/nn/functional.py [hunk 1]
```diff
@@ -2572,31 +2579,51 @@ def gaussian_nll_loss(input, target, var, *, full=False, eps=1e-6, reduction='me
...
-    input = input.view(input.size(0), -1)
-    target = target.view(target.size(0), -1)
     if input.size() != target.size():
         raise ValueError("input and target must have same size")
 
-    # Second dim of var must match that of input or be equal to 1
-    var = var.view(input.size(0), -1)
-    if var.size(1) != input.size(1) and var.size(1) != 1:
-        raise ValueError("var is of incorrect size")
+    # Check var size
```
Comment (albanD) [inline torch/nn/functional.py:2606]: I think you want to update the doc to reflect that.
In particular, var cannot be scalar Tensor anymore? And should be either the same size, have a last dimension of 1 or have one fewer dimension (with all other sizes being the same).

File: torch/nn/functional.py [hunk 2]
```diff
@@ -2611,15 +2638,11 @@ def gaussian_nll_loss(input, target, var, *, full=False, eps=1e-6, reduction='me
     with torch.no_grad():
         var.clamp_(min=eps)
 
-    # Calculate loss (without constant)
-    loss = 0.5 * (torch.log(var) + (input - target)**2 / var).view(input.size(0), -1).sum(dim=1)
-
-    # Add constant to loss term if required
+    # Calculate the loss
+    loss = 0.5 * (torch.log(var) + (input - target)**2 / var)
```
Comment (albanD) [inline torch/nn/functional.py:2642]: This is BC-breaking.
Can you please add a comment in the PR description that explains what was the previous behavior, the new one and why we made that change. As well as a code example on how to recover the old behavior.

cc @jbschlosser This was added fairly recently so I guess we're happy to be BC-breaking here? Maybe we should confirm that in our next sync though?

**Review by albanD — COMMENTED** [review]
Thanks for the update!
I think we want to double check that we're happy with the BC-breaking change.

If so, we should update the tests to make sure that the code behaves as we want for all valid size of `var`.

File: torch/nn/functional.py (hunk 2 above)
Comment (jbschlosser) [reply torch/nn/functional.py:2642]: @albanD Since this loss was just added in 1.8.0, I'm good with the BC-breaking change here to bring the reduction behavior in line with that of other losses.

File: torch/nn/functional.py (hunk 1 above)
Comment (nailimixaM) [reply torch/nn/functional.py:2606]: Yes, thank you - will update the docs.

File: torch/nn/functional.py (hunk 2 above)
Comment (nailimixaM) [reply torch/nn/functional.py:2642]: Thanks both, will update the PR description.

**Comment by albanD** [comment]
btw if you want to look at the rendered doc, you can also check the `pytorch_python_doc_build` and go on the `Artifacts` tab on circle CI. That will contain a full version of the doc compiled with this PR.

File: torch/nn/functional.py (hunk 2 above)
Comment (nailimixaM) [reply torch/nn/functional.py:2642]: I've updated the PR description, let me know if there's anything else to add.

Comment (albanD) [reply torch/nn/functional.py:2638]: The description looks great!

**Review by albanD — APPROVED** [review]
LGTM!

Thanks for updating the doc
